### PR TITLE
Introduce Opera/Sonic based reference implementation of a Processor

### PR DESCRIPTION
**Pull request by HerbertJordan**
This PR introduces a Opera/Sonic derived reference implementation of the Processor interface. 

This reference implementation, registered under the names `geth` and `opera` is intended to server a similar role for the development of Tosca's own Processor implementation as the `geth` interpreter implementation did for the development of the `lfvm` and `evmzero` interpreter.

The introduced Processor implementation has been end-to-end tested by integrating it into Aida and performing a verification run for the first 20M blocks (more to follow). There might still be follow-up bug fixes for this reference implementation, until a full verification run has completed.

Besides the end-to-end tests covered by Aida, a few initial rough Processor shake-down tests have also been included in this PR, covering basic properties. This set should be expanded over time as the semantic of the Processor is worked out in detail.

This PR contains outdated elements of #554 and #555, which should be merged before this PR.

### CODE REVIEW

File: go/vm/geth/processor.go [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+}
+
+func isInternal(transaction vm.Transaction) bool {
+	return transaction.Sender == vm.Address{}
+}
+
+// Source: https://github.com/Fantom-foundation/Sonic/blob/main/opera/contracts/evmwriter/evm_writer.go#L24
+
+// driverAddress is the NodeDriver contract address
+var driverAddress = common.HexToAddress("0xd100a01e00000000000000000000000000000000")
```
Comment (facuMH) [inline go/vm/geth/processor.go:446]: maybe we should consider moving all these global variables and precompiled contracts to another file in the same package.

File: go/vm/geth/processor.go [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+	input []byte,
+	suppliedGas uint64,
+) ([]byte, uint64, error) {
+	if caller != driverAddress {
+		return nil, 0, geth.ErrExecutionReverted
+	}
+	if len(input) < 4 {
+		return nil, 0, geth.ErrExecutionReverted
+	}
+	if bytes.Equal(input[:4], setBalanceMethodID) {
```
Comment (facuMH) [inline go/vm/geth/processor.go:506]: can this be turned into a `switch`  ?

File: go/vm/test/processor_test.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+// Right now, the tested features are minimal. Follow-up work is needed to systematically
+// establish a set of test cases for Processor features.
+//
+// TODO:
+// - test gas price charging
+// - test gas refunding
+// - test left-over gas refunding
+// - test recursive calls
+// - test roll-back on revert
+// - improve test setup
```
Comment (facuMH) [inline go/vm/test/processor_test.go:23]: I guess this comment would cover converting these to TDT.

File: go/vm/test/world_state.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package vm_test
+
+import "github.com/Fantom-foundation/Tosca/go/vm"
+
+type WorldState map[vm.Address]Account
+
+//var _ vm.WorldState = &WorldState{}
```
Comment (facuMH) [inline go/vm/test/world_state.go:17]: commented code ?

File: go/vm/test/world_state.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+type Account struct {
+	Nonce   uint64
+	Balance vm.Value
+	Code    []byte
```
Comment (facuMH) [inline go/vm/test/world_state.go:22]: `vm.Code`

**Review by facuMH — CHANGES_REQUESTED** [review]
Looks good. It's a little hard for me to grasp the whole scope of these changes, but as a base for future implementations is good enough. That said I do think there are a couple improvements that could already be done to help readability.

File: go/vm/geth/geth.go [hunk 6]
```diff
@@ -263,11 +264,25 @@ func (s *stateDbAdapter) CreateContract(common.Address) {
 	// ignored: effect not needed in test environments
 }
 
-func (s *stateDbAdapter) SubBalance(common.Address, *uint256.Int, tracing.BalanceChangeReason) {
-	// ignored: effect not needed in test environments
+func (s *stateDbAdapter) SubBalance(addr common.Address, diff *uint256.Int, _ tracing.BalanceChangeReason) {
```
[the remark this answers is not in the document]
Comment (HerbertJordan) [reply go/vm/geth/geth.go:267]: Everything that is in the `vm` package should be used by the `ct` package. If there are duplicates, the `ct` versions should be removed. Only extra elements that are not needed in the `vm` package should be contributed by the `ct` package.

File: go/vm/test/world_state.go (hunk 4 above)
Comment (HerbertJordan) [reply go/vm/test/world_state.go]: I removed this file from the PR, it was not intended to be there, sorry.

File: go/vm/test/processor_test.go (hunk 3 above)
Comment (HerbertJordan) [reply go/vm/test/processor_test.go:23]: yes

**Comment by HerbertJordan** [comment]
@facuMH the code in the `processor.go` file is copied from its original location in Aida and/or Opera and for being able to back-reference this code, I think we should change as little as possible. Thus, I think I should ignore all style comments on this file, which would lead to a large gap between those implementations, potentially making the identification of errors difficult.

Note: the code of the `geth` processor is right now not intended to be used in any productive system. It is just intended to be our functional reference for building a clean version of it.

**Review by simonlechner — APPROVED** [review]
The changes that are meant to stay look good to me.

**Review by facuMH — APPROVED** [review]

**Review by simonlechner — APPROVED** [review]

**Pull request merged**